lib/elixir/lib/calendar/iso.ex

```diff
@@ -1988,16 +1989,17 @@ defmodule Calendar.ISO do
   end
 
   defp parse_microsecond("." <> rest) do
-    case parse_microsecond(rest, 0, "") do
-      {"", 0, _} ->
+    case parse_microsecond(rest, 0, []) do
+      {[], 0, _} ->
         :error
 
       {microsecond, precision, rest} when precision in 1..6 ->
-        pad = String.duplicate("0", 6 - byte_size(microsecond))
-        {{String.to_integer(microsecond <> pad), precision}, rest}
+        bs = length(microsecond)
+        scale = scale_factor(bs)
+        {{:erlang.list_to_integer(microsecond) * scale, precision}, rest}
 
       {microsecond, _precision, rest} ->
-        {{String.to_integer(binary_part(microsecond, 0, 6)), 6}, rest}
+        {{:erlang.list_to_integer(Enum.take(microsecond, 6)), 6}, rest}
```

Comment:
It can be counted and stopped when we reach 6, the previous case will benefit of that as well because we will know the size